test/Test/Pos/Block/Logic/Mode.hs

```diff
@@ -0,0 +1,296 @@
+    { tpGenUtxo           :: !GenesisUtxo
+    -- ^ Genesis 'Utxo'.
+    , tpSecretKeys        :: !(HashMap StakeholderId SecretKey)
+    -- ^ Secret keys corresponding to 'PubKeyAddress'es from
+    -- genesis 'Utxo'.
+    -- They are stored in map (with 'StakeholderId' as key) to make it easy
+    -- to find 'SecretKey' corresponding to given 'StakeholderId'.
+    -- In tests we often want to have inverse of 'hash' and 'toPublic'.
+    , tpStakeDistribution :: !StakeDistribution
+    -- ^ Stake distribution which was used to generate genesis utxo.
+    -- It's primarily needed to see which distribution was used (e. g.
+    -- when test fails).
+    }
+
+instance Buildable TestParams where
+    build TestParams {..} =
+        bprint ("TestParams {\n"%
+                "  utxo = "%utxoF%"\n"%
+                "  secret keys: "%int%" items\n"%
+                "  stake distribution: "%shown%"\n"%
+                "  stakeholders: "%listJson%"\n"%
+                "}\n")
+            utxo
+            (length tpSecretKeys)
+            tpStakeDistribution
+            (HM.keys tpSecretKeys)
+      where
+        utxo = tpGenUtxo & \(GenesisUtxo u) -> u
+
+instance Show TestParams where
+    show = formatToString build
+
+-- More distributions can be added if we want (e. g. RichPoor).
+genSuitableStakeDistribution :: Word -> Gen StakeDistribution
+genSuitableStakeDistribution stakeholdersNum =
+    oneof [genFlat{-, genBitcoin-}, pure ExponentialStakes]
+  where
+    totalCoins = mkCoin <$> choose (fromIntegral stakeholdersNum, unsafeGetCoin maxBound)
+    genFlat =
+        FlatStakes stakeholdersNum <$> totalCoins
+    -- Apparently bitcoin distribution is broken (it produces total stake
+    -- greater than requested one) and I don't think it's worth fixing it.
+    -- genBitcoin = BitcoinStakes stakeholdersNum <$> totalCoins
+
+instance Arbitrary TestParams where
+    arbitrary = do
+        secretKeysList <- toList @(NonEmpty SecretKey) <$> arbitrary -- might have repetitions
+        let toSecretPair sk = (addressHash (toPublic sk), sk)
+        let tpSecretKeys = HM.fromList $ map toSecretPair secretKeysList
+        tpStakeDistribution <-
+            genSuitableStakeDistribution (fromIntegral $ length tpSecretKeys)
+        let zipF secretKey (coin, toaDistr) =
+                let addr = makePubKeyAddress (toPublic secretKey)
+                    toaOut = TxOut addr coin
+                in (TxIn (unsafeHash addr) 0, TxOutAux {..})
+        let tpGenUtxo =
+                GenesisUtxo . M.fromList $
+                zipWith
+                    zipF
+                    (toList tpSecretKeys)
+                    (stakeDistribution tpStakeDistribution)
+        return TestParams {..}
+
+----------------------------------------------------------------------------
+-- Main context
+----------------------------------------------------------------------------
+
+modeContext [d|
+    data BlockTestContext = BlockTestContext
```

Comment:
I think you can use `RealMode`. Why do you need this context? It will be changed to use pure db after Misha implements pure DB, I think. And it doesn't make sense to use pure DB in generator.